flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+ * it only cumulates data when they span multiple input buffers. For the buffer part, it
+ * copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class ZeroCopyNettyMessageDecoder extends ChannelInboundHandlerAdapter {
+
+    private static final int INITIAL_MESSAGE_HEADER_BUFFER_LENGTH = 128;
+
+    /** The parser to parse the message header. */
+    private final NettyMessageParser messageParser;
+
+    /** The buffer used to cumulate the frame header part. */
+    private ByteBuf frameHeaderBuffer;
+
+    /** The buffer used to receive the message header part. */
+    private ByteBuf messageHeaderBuffer;
+
+    /** Which part of the current message is being decoded. */
+    private DecodeStep decodeStep;
+
+    /** How many bytes have been decoded in current step. */
+    private int decodedBytesOfCurrentStep;
+
+    /** The intermediate state when decoding the current message. */
+    private final MessageDecodeIntermediateState intermediateState;
+
+    ZeroCopyNettyMessageDecoder(NettyMessageParser messageParser) {
+        this.messageParser = messageParser;
+        this.intermediateState = new MessageDecodeIntermediateState();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        frameHeaderBuffer = ctx.alloc().directBuffer(NettyMessage.FRAME_HEADER_LENGTH);
+        messageHeaderBuffer = ctx.alloc().directBuffer(INITIAL_MESSAGE_HEADER_BUFFER_LENGTH);
+
+        decodeStep = DecodeStep.DECODING_FRAME_HEADER;
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+        if (intermediateState.messageHeaderParseResult != null) {
+            Buffer buffer = intermediateState.messageHeaderParseResult.getTargetDataBuffer();
+
+            if (buffer != null) {
+                buffer.recycleBuffer();
+            }
+        }
+
+        clearState();
+
+        frameHeaderBuffer.release();
+        messageHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.readableBytes() > 0) {
+                if (decodeStep == DecodeStep.DECODING_FRAME_HEADER) {
+                    ByteBuf toDecode = cumulateBufferIfNeeded(frameHeaderBuffer, data, FRAME_HEADER_LENGTH);
+
+                    if (toDecode != null) {
+                        decodeFrameHeader(toDecode);
+
+                        decodedBytesOfCurrentStep = 0;
```

Comment:
This is reductant as commented in https://github.com/apache/flink/pull/7368/files#r376207885

Reply:
Have removed this line